client/client_test.go

```diff
@@ -6495,6 +6496,272 @@ func testExportAttestations(t *testing.T, sb integration.Sandbox) {
+		res := gateway.NewResult()
+
+		st := llb.Image("busybox")
+		def, err := st.Marshal(sb.Context())
+		require.NoError(t, err)
+
+		r, err := c.Solve(ctx, gateway.SolveRequest{
+			Definition: def.ToPB(),
+		})
+		if err != nil {
+			return nil, err
+		}
+		ref, err := r.SingleRef()
+		if err != nil {
+			return nil, err
+		}
+		_, err = ref.ToState()
+		if err != nil {
+			return nil, err
+		}
+		res.AddRef(pk, ref)
+
+		expPlatforms := &exptypes.Platforms{
+			Platforms: []exptypes.Platform{{ID: pk, Platform: p}},
+		}
+		dt, err := json.Marshal(expPlatforms)
+		if err != nil {
+			return nil, err
+		}
+		res.AddMeta(exptypes.ExporterPlatformsKey, dt)
+
+		var img ocispecs.Image
+		img.Config.Cmd = []string{"/bin/sh", "-c", `for f in $BUILDKIT_SCAN_SOURCES/*; do echo "{\"success\": false}" > $BUILDKIT_SCAN_DESTINATIONS/$(basename $f)/spdx.json; done`}
+		config, err := json.Marshal(img)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to marshal image config")
+		}
+		res.AddMeta(fmt.Sprintf("%s/%s", exptypes.ExporterImageConfigKey, pk), config)
+
+		return res, nil
+	}
+
+	scannerTarget := registry + "/buildkit/testsbomscanner:latest"
+	_, err = c.Build(sb.Context(), SolveOpt{
+		Exports: []ExportEntry{
+			{
+				Type: ExporterImage,
+				Attrs: map[string]string{
+					"name": scannerTarget,
+					"push": "true",
+				},
+			},
+		},
+	}, "", scannerFrontend, nil)
+	require.NoError(t, err)
+
+	makeTargetFrontend := func(attest bool) func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+		return func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+			res := gateway.NewResult()
+
+			// build image
+			st := llb.Scratch().File(
+				llb.Mkfile("/greeting", 0600, []byte("hello world!")),
+			)
+			def, err := st.Marshal(ctx)
+			if err != nil {
+				return nil, err
+			}
+			r, err := c.Solve(ctx, gateway.SolveRequest{
+				Definition: def.ToPB(),
+			})
+			if err != nil {
+				return nil, err
+			}
+			ref, err := r.SingleRef()
+			if err != nil {
+				return nil, err
+			}
+			_, err = ref.ToState()
+			if err != nil {
+				return nil, err
+			}
+			res.AddRef(pk, ref)
+
+			expPlatforms := &exptypes.Platforms{
+				Platforms: []exptypes.Platform{{ID: pk, Platform: p}},
+			}
+			dt, err := json.Marshal(expPlatforms)
+			if err != nil {
+				return nil, err
+			}
+			res.AddMeta(exptypes.ExporterPlatformsKey, dt)
+
+			// build attestations
+			if attest {
+				st = llb.Scratch().
+					File(llb.Mkfile("/result.spdx", 0600, []byte(`{"success": true}`)))
+				def, err = st.Marshal(ctx)
+				if err != nil {
+					return nil, err
+				}
+				r, err = c.Solve(ctx, gateway.SolveRequest{
+					Definition: def.ToPB(),
+				})
+				if err != nil {
+					return nil, err
+				}
+				refAttest, err := r.SingleRef()
+				if err != nil {
+					return nil, err
+				}
+				_, err = ref.ToState()
+				if err != nil {
+					return nil, err
+				}
+
+				res.AddAttestation(pk, &result.InTotoAttestation{
```

Comment:
This should be done is separate PR but this call looks super hacky and needs to be cleaned up. It should read just:

```
res.AddAttestation(pk, &result.InTotoAttestation{
					PredicatePath:   "/result.spdx",
					PredicateType:   intoto.PredicateSPDX,
				}, refAttest)
```

Reply:
I originally had:

```
res.AddInTotoAttestation(pk, &result.InTotoAttestation{
					PredicatePath:   "/result.spdx",
					PredicateType:   intoto.PredicateSPDX,
				}, refAttest)
```

The reason we can't easily have a general `AddAttestation` with a single Ref argument is that I don't think we can necessarily assume 1 ref per attestation.

For example, if we were attesting about a file (e.g. a built binary, with it's own sbom for dependencies or something), we would have a `PredicateRefKey` and a `FileSubjectRefKey` (with a `FileSubjectRefPath` or similar). In this case, we'd need to take 2 arguments, and I don't see how we could easily take variadic arguments, since the order isn't completely clear.

The map solution isn't pretty, but it does solve this problem - though I'd love some alternatives. Maybe we could just enforce the order per-different attestation type?

Reply:
What is `FileSubjectRefKey`. We should remove the things we don't have a use case for. Just leave type field check so that changes are backward compatible.

If there is any use case for these attestations with subjects not managed by buildkit then these are completely user's responsibility. We will not try to invent complicated protocols so that user can avoid calling sha256 on their own and instead somehow buildkit would compute a hash for a special arrangement of bytes.

Reply:
Addressed in https://github.com/moby/buildkit/pull/3073.